debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlConnectorIT.java

```diff
@@ -1983,4 +2014,57 @@ private void waitForStreamingRunning(String serverName) throws InterruptedExcept
         return uc != null ? uc : records.recordsForTopic(RO_DATABASE.topicForTable("products"));
     }
+
+    @Test
+    @FixFor("DBZ-1531")
+    public void shouldEmitOldkeyHeaderOnPrimaryKeyUpdate() throws Exception {
+        config = DATABASE.defaultConfig()
+                .with(MySqlConnectorConfig.SNAPSHOT_MODE, MySqlConnectorConfig.SnapshotMode.NEVER)
+                .build();
+
+        // Start the connector ...
+        start(MySqlConnector.class, config);
+
+        // ---------------------------------------------------------------------------------------------------------------
+        // Consume all of the events due to startup and initialization of the database
+        // ---------------------------------------------------------------------------------------------------------------
+        SourceRecords records = consumeRecordsByTopic(INITIAL_EVENT_COUNT); // 6 DDL changes
+        assertThat(records.recordsForTopic(DATABASE.topicForTable("orders")).size()).isEqualTo(5);
+
+        try (MySQLConnection db = MySQLConnection.forTestDatabase(DATABASE.getDatabaseName());) {
+            try (JdbcConnection connection = db.connect()) {
+                connection.execute("UPDATE orders SET order_number=10303 WHERE order_number=10003");
+            }
+        }
+        // Consume the update of the PK, which is one insert followed by a delete followed by a tombstone ...
+        records = consumeRecordsByTopic(3);
+        List<SourceRecord> updates = records.recordsForTopic(DATABASE.topicForTable("orders"));
+
+        assertThat(updates.size()).isEqualTo(3);
+
+        SourceRecord deleteRecord = updates.get(0);
+        assertEquals(1, deleteRecord.headers().size()); // to be removed/updated once we set additional headers
+        Header keyPKUpdateHeader = getPKUpdateNewKeyHeader(deleteRecord);
+        assertEquals(Integer.valueOf(10303), ((Struct) keyPKUpdateHeader.value()).getInt32("order_number"));
+
+        SourceRecord insertRecord = updates.get(2);
+        assertEquals(1, insertRecord.headers().size()); // to be removed/updated once we set additional headers
+        keyPKUpdateHeader = getPKUpdateOldKeyHeader(insertRecord);
+        assertEquals(Integer.valueOf(10003), ((Struct) keyPKUpdateHeader.value()).getInt32("order_number"));
+
+        try (MySQLConnection db = MySQLConnection.forTestDatabase(DATABASE.getDatabaseName());) {
+            try (JdbcConnection connection = db.connect()) {
+                connection.execute("UPDATE orders SET quantity=5 WHERE order_number=10004");
+            }
+        }
+        records = consumeRecordsByTopic(5);
+        updates = records.recordsForTopic(DATABASE.topicForTable("orders"));
+        assertThat(updates.size()).isEqualTo(1);
+
+        SourceRecord updateRecord = updates.get(0);
+        assertEquals(0, updateRecord.headers().size()); // to be removed/updated once we set additional headers
```

Comment:
Ok, so that's still WIP, right :)

Btw. IIRC you mentioned via DM that it couldn't be done for MySQL because it would send DELETE + UPDATE by itself. If true indeed, it seems there should be no test for MySQL to begin with?

Reply:
Did you read my previous comment? :thinking:

Reply:
Ah, I see, I mixed this up then. Thanks for clarifying.